File: lib/domain/unknown.ml
type reason =
  | Unsupported_syntax of string
  | External_state of string
  | Unresolved_dependency of string
  | Recursive_call of string
  | Dynamic_string of string
  | Phase_unavailable of string
  | Missing_evidence of string
  | Resource_limit of string

let compare = Stdlib.compare

let kind_and_detail = function
  | Unsupported_syntax detail -> ("unsupported_syntax", detail)
  | External_state detail -> ("external_state", detail)
  | Unresolved_dependency detail -> ("unresolved_dependency", detail)
  | Recursive_call detail -> ("recursive_call", detail)
  | Dynamic_string detail -> ("dynamic_string", detail)
  | Phase_unavailable detail -> ("phase_unavailable", detail)
  | Missing_evidence detail -> ("missing_evidence", detail)
  | Resource_limit detail -> ("resource_limit", detail)

let to_string reason =
  let kind, detail = kind_and_detail reason in
  if detail = "" then kind else kind ^ ": " ^ detail

let to_json reason =
  let kind, detail = kind_and_detail reason in
  Json.Object [ ("detail", Json.String detail); ("kind", Json.String kind) ]

File: src/domain/unknown.rs
use crate::foundation::JsonValue;
use std::collections::BTreeMap;
use std::fmt;

#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub enum UnknownReason {
    UnsupportedSyntax(String),
    ExternalState(String),
    UnresolvedDependency(String),
    RecursiveCall(String),
    DynamicString(String),
    PhaseUnavailable(String),
    MissingEvidence(String),
    ResourceLimit(String),
}

impl UnknownReason {
    #[must_use]
    pub fn kind_and_detail(&self) -> (&'static str, &str) {
        match self {
            Self::UnsupportedSyntax(detail) => ("unsupported_syntax", detail),
            Self::ExternalState(detail) => ("external_state", detail),
            Self::UnresolvedDependency(detail) => ("unresolved_dependency", detail),
            Self::RecursiveCall(detail) => ("recursive_call", detail),
            Self::DynamicString(detail) => ("dynamic_string", detail),
            Self::PhaseUnavailable(detail) => ("phase_unavailable", detail),
            Self::MissingEvidence(detail) => ("missing_evidence", detail),
            Self::ResourceLimit(detail) => ("resource_limit", detail),
        }
    }

    #[must_use]
    pub fn to_json(&self) -> JsonValue {
        let (kind, detail) = self.kind_and_detail();
        JsonValue::Object(BTreeMap::from([
            ("detail".to_owned(), JsonValue::String(detail.to_owned())),
            ("kind".to_owned(), JsonValue::String(kind.to_owned())),
        ]))
    }
}

impl fmt::Display for UnknownReason {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let (kind, detail) = self.kind_and_detail();
        if detail.is_empty() {
            formatter.write_str(kind)
        } else {
            write!(formatter, "{kind}: {detail}")
        }
    }
}
